File: code/agents/InsightAgent.java
package agents;

import com.google.cloud.bigquery.FieldValue;
import com.google.cloud.bigquery.FieldValueList;
import com.google.cloud.bigquery.TableResult;

import java.util.ArrayList;
import java.util.List;

/**
 * InsightAgent autonomously analyzes query results to derive meaningful insights.
 * Usage: Instantiate and call analyze(TableResult).
 */
public class InsightAgent {

    public InsightAgent() {
        // Initialize any resources if needed
    }

    // Example method to analyze trends or perform simple insights on query results
    public List<String> analyze(TableResult queryResult) {
        List<String> insights = new ArrayList<>();

        // Simple example: count number of rows and provide insight
        long rowCount = queryResult.getTotalRows();
        insights.add("Total rows in dataset: " + rowCount);

        // Example: analyze first column values (if numeric)
        // This is just a placeholder for actual analysis logic
        double sum = 0;
        long count = 0;

        for (FieldValueList row : queryResult.iterateAll()) {
            FieldValue val = row.get(0);
            if (val.getValue() instanceof Number) {
                sum += ((Number) val.getValue()).doubleValue();
                count++;
            }
        }

        if (count > 0) {
            double avg = sum / count;
            insights.add("Average value of first column: " + avg);
        } else {
            insights.add("No numeric data found in first column for analysis.");
        }

        return insights;
    }

    /**
     * Standalone entry point for running the InsightAgent independently (demo only).
     */
    public static void main(String[] args) {
        System.out.println("InsightAgent: Please use this agent with a TableResult from QueryAgent.");
    }
}

File: code/agents/insight_agent.py
from shared.dataset import Dataset

class InsightAgent:
    def __init__(self):
        pass

    def analyze(self, dataset):
        insights = []
        
        if not dataset:
            return ["No dataset to analyze."]

        row_count = dataset.get_total_rows()
        insights.append(f"Total rows in dataset: {row_count}")

        sum_val = 0
        count = 0
        
        for row in dataset.get_rows():
            if len(row) > 1:
                val = row[1]
                if isinstance(val, (int, float)):
                    sum_val += val
                    count += 1
                elif isinstance(val, str) and val.replace('.', '', 1).isdigit():
                     sum_val += float(val)
                     count += 1

        if count > 0:
            avg = sum_val / count
            insights.append(f"Average value of second column: {avg}")
        else:
            insights.append("No numeric data found in second column for analysis.")

        return insights

if __name__ == "__main__":
    print("InsightAgent: Please use this agent with a Dataset from QueryAgent.")
